### ISSUE TITLE
Add dimensionality to variable listing

### ISSUE THREAD

**Issue by zklaus** [clix-meta/clix-meta#28]
In #20, @aulemahal suggest adding dimension information to the variables:

> `Input.dimensionality`, this one is more tricky. Xclim checks the input units, but is not strict, so it needs the dimensionality of the variable (i.e. [temperature] or [length]). Because of xclim-specific additions like "[precipitation]" I doubt we can include this directly, but what about a canonical_units field? That would fit the cf standard names table.

It is certainly a good idea to keep track of the units of input data. In a way, this is redundant to add, since the `standard_name` already implies canonical units via CF.

However, we too are dealing with precipitation in a special way since this quantity is commonly given in flux or cumulative units.

@aulemahal and @larsbarring, is precipitation the only phenomenon that requires this kind of special treatment, or do other variables also deviate from the CF canonical units?

The description of precipitation in CF is, ahem, rich. The standard names fall in the three categories fluxes, rates, amounts (there are also mass fractions, but I think we can disregard that for index calculation purposes).

If we think that precipitation is the only phenomenon that needs this kind of special treatment, I think we should have a special issue for that and otherwise deal with the dimensionality simply by referring to the canonical units implied by the respective `standard_name`.

Note that precipitation may come in liquid or various solid forms.

**Comment by larsbarring** [comment]
Yes, for precipitation the dimensionality is indeed important. But I am not sure that the concept of canonical unit make much sense without a standard name (or an equally well-defined replacement). There is a 1-to-1 relationship between standard name and canonical unit. If you on the other hand specify "precipitation" or "rainfall or "snowfall" as input the user might be free to provide data in any [reasonable] unit which then with some manipulation should possible to link to a standard name. If not it might be a good indication that the provided unit is not reasonable. 

To give an example of what might happen if one relaxes the link between standard name and canonical unit, or simply omit the standard name, is given in [CF Example 7.14](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#monthly-max-daily-precip-totals-ex), where *Monthly-maximum daily precipitation totals* is given as a mass (`kg`), i.e. integrated over time and over some area (possibly gridcell, or raingauge orifice).

Right now no other concrete example comes to mind, but in principle we face the same issue when it comes to evapo[transpi]ration, and possibly other 'flows of water'.

**Comment by larsbarring** [comment]
Following up on this: it might be more appropriate to specify what is allowable in the *variable sheet*cf. #27 

And, as a side note, the `standard_name` and `canonical_unit` may not always give a complete and exhaustive specification of a variable, as for example with [temperature](https://github.com/cf-convention/discuss/issues/101)

**Comment by aulemahal** [comment]
Indeed, the standard names table already provide canonical units so no need to rewrite them in the tables here. 
Regarding the "relaxing" of input units I think this relates to another issue I raised in #20:

> The [thresholded_]statistics index functions can reduce with a sum. Ex. the ss indicator reduces sund, given in seconds, with > a sum over days within the period, output is in hours. All good.
> However, the rr indicator reduces pr and the output of the sum is supposed to be in [mm]. While I understand that this works > in the case where pr is in mm/d, there is a implicit conversion from flux to amount here. and the canonical unit of pr is indeed > kg m-2 s-1, which will not be equivalent to mm when summed over days. Should there be a way to indicate this in the table or  > in the index functions? Like adding a flag for indicating a rate-to-amount conversion?

I.e.: My view would be to avoid any implicit transformations. For example, the CF example 7.14 you linked seems to me faulty because the input was precipitation (kg m-2 s-1) and the cell methods show  "sum within days" and "maximum over days", this gives [kg m-2] at best, interpreting "sum" by "integral". Where was the area integral?
So, I see 2 "solutions" here, and prefer the 1st by far:

1. Add the necessary "flags" in the tables so that all the dimensional analysis from the inputs to the output canonical units are explicit. 
2. Assume specific implicit conversions (like [kg m-2 s-1]  to [mm/day]  (when it is precipitation, using 1000 kg/m³) and then to [mm] (assuming daily input)). 

In this idea, if input data is not given in the canonical units (like precip in mm/d), it is the responsibility of the user/computing library to make the conversion. Not the responsibility of clix-meta!

**Comment by zklaus** [comment]
See also comments [here](https://github.com/clix-meta/clix-meta/issues/20#issuecomment-824248199)
